### PR TITLE
systemd: switch to unified cgroup hierarchy by default

### LINKED ISSUE #73800: Migrate to cgroupsv2
https://github.com/NixOS/nixpkgs/pull/68096#issuecomment-527826574

For the uninitiated https://www.kernel.org/doc/Documentation/cgroup-v2.txt , see also section R stating the rationales for v2.

This is not an immediately actionable issue, the first major blocker seems to be waiting for docker to migrate.

Some more keywords for searchability (maybe): cgroups cgroups-v2

Currently AFAICT all cgroup controllers are acquired by the v1 system, because a controller can only be mounted (or whatever the word is) on one cgroup hierarchy at a time.

https://web.archive.org/web/20191120061819/https://medium.com/nttlabs/cgroup-v2-596d035be4d7 suggests 

```
Docker / Moby will gain the support for cgroup v2, as soon as runc
and containerd gains the support. 
```

```
Docker/Moby+containerd+runc will follow soon. If everything goes well,
we might be able to get nightly binaries for cgroup v2 by the end of 2019.
```

cc @arianvp

### PR DESCRIPTION
**Pull request by flokli**
Other distros, like Fedora 32 have already switched to it.

Users that need to keep using the old cgroup hierarchy (namely because
they want to run docker, and not some of the alternatives supporting
cgroupsv2 can re-enable it by setting `systemd.enableUnifiedCgroupHierarchy = false;`.

Fixes #73800

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/doc/manual/release-notes/rl-2103.xml
Comment (grahamc) [inline nixos/doc/manual/release-notes/rl-2103.xml:272]: I'm not sure we need to appeal to other distros as an "authority".  How about something like:

```suggestion
     NixOS now defaults to the unified cgroup hierarchy (cgroupsv2).
     See the <link xlink:href="https://www.redhat.com/sysadmin/fedora-31-control-group-v2">Fedora Article for 32</link>
     for details on why this is desirable, and how it impacts containers.
```

**Review by grahamc — COMMENTED** [review]
I have this in my config:

```
   boot.kernelParams = (if config.virtualisation.docker.enable then [] else [ "cgroup_no_v1=all" "systemd.unified_cgroup_hierarchy=yes" ]);
```

maybe we should automatically specify this for the docker services?

File: pkgs/os-specific/linux/systemd/default.nix
Comment (andir) [inline pkgs/os-specific/linux/systemd/default.nix:286]: Ugh, why not be explicit? Both here and via the module system. If we have this in the module system we can throw proper warnings/errors when docker is enabled.

Comment (flokli) [reply pkgs/os-specific/linux/systemd/default.nix:286]: Makes sense, done.

File: nixos/doc/manual/release-notes/rl-2103.xml
Comment (flokli) [reply nixos/doc/manual/release-notes/rl-2103.xml]: Updated, thanks!

**Review by flokli — COMMENTED** [review]
> I have this in my config: 
```
boot.kernelParams = (if config.virtualisation.docker.enable then [] else [ "cgroup_no_v1=all" "systemd.unified_cgroup_hierarchy=yes" ]);
```

> maybe we should automatically specify this for the docker services?

Yes, adding `"systemd.unified_cgroup_hierarchy=0` if docker is enabled might make sense, at least as long as docker doesn't support cgroupv2.
We probably should add link to the upstream tracking issues.

It seems it's supported in runc: https://github.com/opencontainers/runc/issues/2315#issuecomment-700423212 and should be supported in docker 20, which was "not officially released yet, and its ETA still unpublished" in September 2020: https://github.com/moby/moby/issues/40360#issuecomment-699867043

**Comment by SuperSandro2000** [comment]
Docker 20.10.0-rc1 was released today which should add cgroupsv2 and apparmor 3.0-beta support. Maybe we can wait until it is released?

**Comment by flokli** [comment]
@SuperSandro2000 are you aware of any timelines on 20.* getting released? Would you be up to packaging their rc, to see how it behaves? ;-)

I surely want to add/extend some more tests about this before merging, and see how config options for other container runtimes might need to look like (if it requires special configuration).

**Comment by ajs124** [comment]
This will make user@ services fail when using with the hardened profile. Specifically [this line of the hardened profile](https://github.com/NixOS/nixpkgs/blob/b6e21a7/nixos/modules/profiles/hardened.nix#L20) and [this issue](https://github.com/systemd/systemd/issues/12955#issuecomment-508490893) marked as "not supported".

There are several solutions to this and it definitely should not block this, but I thought it was worth noting.

cc @joachifm (because of the hardened profile).

**Comment by flokli** [comment]
I also did some more experiments:

 - cri-o works fine with unified cgroups.
 - k3s with docker as container runtime obviously still needs unified cgroups disabled
 - k3s with their bundled containerd doesn't work with cgroupsv2 either

I made sure the unified cgroup hierarchy is disabled in these cases and the tests succeed again.

**Comment by flokli** [comment]
I added some more tests on the accounting parts. I couldn't reliably get IO Accounting on individual units tested, but the test ensures `system.slice` has non-zero IO accounting values :-)

This is ready for review, PTAL.

**Comment by SuperSandro2000** [comment]
> are you aware of any timelines on docker 20.* getting released?

No but I assume it could take a month.

> Would you be up to packaging their rc, to see how it behaves?

I can get it to build but for testing I miss experience.

**Comment by flokli** [comment]
> > are you aware of any timelines on docker 20.* getting released?
> 
> No but I assume it could take a month.

Okay. So I propose we keep this PR as-is (disable unified cgroup when you enable docker), and we can remove this once docker has landed and supports it.

> > Would you be up to packaging their rc, to see how it behaves?
> 
> I can get it to build but for testing I miss experience.

`nixosTests.docker` should probably be enough to verify basic functionality, and once that works, people could take a more closer look in the PR :-)

**Comment by SuperSandro2000** [comment]
> nixosTests.docker should probably be enough to verify basic functionality, and once that works, people could take a more closer look in that new PR :-)

KVM does not want to work with nix tests on my machine. Not sure why.

File: nixos/tests/systemd.nix
Comment (ajs124) [inline nixos/tests/systemd.nix:183]: Are you sure this won't break?

**Review by ajs124 — APPROVED** [review]
LGTM, in general. Some notes, although they're not too important.

File: nixos/modules/security/hidepid.nix
Comment (flokli) [reply nixos/modules/security/hidepid.nix:29]: For now, I just disabled it where it caused breakage, so the test kept working without any config change. I'm happy to change things here, but don't think it needs to block this PR.

File: nixos/modules/services/cluster/k3s/default.nix
Comment (flokli) [reply nixos/modules/services/cluster/k3s/default.nix:82]: This currently doesn't work with k3s, and I'd expect people enabling k3s to want to have it working.

File: nixos/tests/systemd.nix
Comment (flokli) [reply nixos/tests/systemd.nix:183]: It was constant here at least.

File: nixos/modules/system/boot/systemd.nix
Comment (arianvp) [inline nixos/modules/system/boot/systemd.nix:553]: *bikeshed*: Can we call this `enableUnifiedCgroupHierarchy`  so it's the same naming scheme as the other boolean options in this module like `enableCgroupAccounting` ?

Comment (flokli) [reply nixos/modules/system/boot/systemd.nix]: bikeshed accepted and updated.

File: nixos/tests/podman.nix
Comment (zowoq) [inline nixos/tests/podman.nix:60]: Can we keep the explicit `--runtime=crun`/ `Run container rootless with crun`  please?

Comment (flokli) [reply nixos/tests/podman.nix:60]: Can you elaborate on why? `podman` seems to default to it with cgroupsv2, and if we view this test from a user perspective, they just want to `podman run` as non-root, without explicitly configuring a specific runtime.

also cc @adisbladis

Comment (adisbladis) [reply nixos/tests/podman.nix:60]: I just want to say I agree with @flokli on this one. It's better that we test the users default config.
Also I seem to recall seeing some intent to switch back to `runc` by default at some point once it's gained the required cgroupsv2 support (don't quote me on this one).

Comment (zowoq) [reply nixos/tests/podman.nix:60]: Can we do this instead?

```diff
diff --git a/nixos/tests/podman.nix b/nixos/tests/podman.nix
index cd8c2b4308c..74581f5c847 100644
--- a/nixos/tests/podman.nix
+++ b/nixos/tests/podman.nix
@@ -55,13 +55,13 @@ import ./make-test-python.nix (
 
       with subtest("Run container rootless with runc"):
           podman.succeed(su_cmd("tar cv --files-from /dev/null | podman import - scratchimg"))
-          podman.succeed(
+          podman.fail(
               su_cmd(
                   "podman run --runtime=runc -d --name=sleeping -v /nix/store:/nix/store -v /run/current-system/sw/bin:/bin scratchimg /bin/sleep 10"
               )
           )
-          podman.succeed(su_cmd("podman ps | grep sleeping"))
-          podman.succeed(su_cmd("podman stop sleeping"))
+          podman.fail(su_cmd("podman ps | grep sleeping"))
+          podman.fail(su_cmd("podman stop sleeping"))
           podman.succeed(su_cmd("podman rm sleeping"))
 
       with subtest("Run container rootless with crun"):

```

Comment (flokli) [reply nixos/tests/podman.nix:60]: I don't think our test should suddenly fail if podmans runc backend starts working in rootless scenarios again.

I'm fine adding a test for the default backend in as-root scenarios, but made the rootless runc subtest a comment, which we can re-add once it gets implemented.

Comment (zowoq) [reply nixos/tests/podman.nix:68]: > I don't think our test should suddenly fail if podmans runc backend starts working in rootless scenarios again.
> 
> I'm fine adding a test for the default backend in as-root scenarios, but made the rootless runc subtest a comment, which we can re-add once it gets implemented.

All of these packages have `passthru.tests` so the failure will be noticed when it does start working.

Can you move the default tests after the explict runtime tests please? If it does fail because of a runtime problem having it fail on the explict test would be clearer.

Comment (flokli) [reply nixos/tests/podman.nix:68]: I can move the test order, sure, but I still don't think the test should suddenly fail if *more* things work. `podman` is still functional if a backend that was known to be broken starts working again. We shouldn't then fail the test.

Comment (flokli) [reply nixos/tests/podman.nix:68]: I'd expect the maintainers of this package to read the release notes while bumping, and updating/extending the test accordingly once new features are added.

File: nixos/modules/virtualisation/docker.nix
Comment (andir) [inline nixos/modules/virtualisation/docker.nix:159]: Would it make sense to make this conditional on the docker version? Users can set the docker package at runtime and defaulting this to the right version automatically might be a nice perk that we can implement rather quickly in NixOS?

Comment (flokli) [reply nixos/modules/virtualisation/docker.nix:159]: We don't know yet if docker 20.10 really works. I'm happy to take a closer look at this once it is in Nixpkgs. There currently is no docker working with enabled unified cgroup hierarchy, so I'm not much in favor of code for the eventuality of it being there ;-)

**Comment by flokli** [comment]
Let's get this in. If there's other applications that don't really work with cgroupsv2, disabling it for them is pretty trivial by setting `systemd.enableUnifiedCgroupHierarchy = false;`.

**Pull request merged**
